proxy/context.go

```diff
@@ -44,84 +44,68 @@ type Environment struct {
 	Mesh *proxyconfig.ProxyMeshConfig
 }
 
-// Role declares the proxy node role in the mesh
-type Role interface {
-	// nolint: megacheck
-	isProxyRole()
+// Role defines the proxy role
+type Role struct {
+	// Type specifies the role type
+	Type RoleType
 
-	// ServiceNode encodes the role information into a string
-	ServiceNode() string
-}
-
-// Sidecar defines the sidecar proxy role
-type Sidecar struct {
-	// IPAddress is the IP address of the proxy used to identify it and its
-	// co-located service instances. Example: "10.60.1.6"
+	// IPAddress is the IP address of the proxy used to identify its co-located
+	// service instances. Example: "10.60.1.6"
 	IPAddress string
 
 	// ID is the unique platform-specific sidecar proxy ID
 	ID string
 
-	// Domain defines the DNS domain suffix for short hostnames
+	// Domain defines the DNS domain suffix for short hostnames (e.g.
+	// "default.svc.cluster.local")
 	Domain string
 }
 
-func (Sidecar) isProxyRole() {}
+// RoleType decides what role the proxy serves in the mesh
+type RoleType string
+
+const (
+	// Sidecar type is used for sidecar proxies in the application containers
+	Sidecar RoleType = "sidecar"
+
+	// Ingress type is used for cluster ingress proxies
+	Ingress RoleType = "ingress"
+
+	// Egress type is used for cluster egress proxies
+	Egress RoleType = "egress"
+)
+
+// ServiceNode encodes the role into a URI-acceptable string
+func (role Role) ServiceNode() string {
+	return strings.Join([]string{
+		string(role.Type), role.IPAddress, role.ID, role.Domain,
+	}, serviceNodeSeparator)
 
-// ServiceNode for sidecar
-func (role Sidecar) ServiceNode() string {
-	return fmt.Sprintf("%s|%s|%s", role.IPAddress, role.ID, role.Domain)
 }
 
-// DecodeServiceNode is the inverse of sidecar service node
-func DecodeServiceNode(s string) (Sidecar, error) {
-	parts := strings.Split(s, "|")
-	out := Sidecar{}
+// ParseServiceNode is the inverse of service node function
+func ParseServiceNode(s string) (Role, error) {
+	parts := strings.Split(s, serviceNodeSeparator)
+	out := Role{}
 
-	if len(parts) > 0 {
-		out.IPAddress = parts[0]
-	}
-	if len(parts) > 1 {
-		out.ID = parts[1]
+	if len(parts) != 4 {
+		return out, errors.New("missing parts in the role service node")
 	}
 
-	if len(parts) > 2 {
-		out.Domain = parts[2]
-	}
+	out.Type = RoleType(parts[0])
+	out.IPAddress = parts[1]
+	out.ID = parts[2]
+	out.Domain = parts[3]
 	return out, nil
 }
 
 const (
-	// EgressNode is the service node for egress proxies
-	EgressNode = "egress"
-
-	// IngressNode is the service node for ingress proxies
-	IngressNode = "ingress"
+	serviceNodeSeparator = "~"
```

Comment:
only ":" messes with statsd in envoy, other characters are fine. this also has to be a legit URI character.